pkg/kv/kvserver/store.go

```diff
@@ -411,8 +411,277 @@ func (rs *storeReplicaVisitor) EstimatedCount() int {
+Ranges own a contiguous chunk of addressable keyspace, where the word
+"addressable" is a fine-print that interested readers can learn about in
+keys.Addr; in short the Replica data is logically contiguous, but not
+contiguous when viewed as Engine kv pairs.
+
+Considerable complexity is incurred by the features of dynamic re-sharding
+and relocation, i.e. range splits, range merges, and range relocation (also
+called replication changes, or, in the case of lateral movement,
+rebalancing). Each of these interact heavily with the Range a consensus group
+(of which each Replica is a member). All of these intricacies are described
+at a high level in this comment.
+
+RangeDescriptor
+
+A roachpb.RangeDescriptor is the configuration of a Range. It is an
+MVCC-backed key-value pair (where the key is derived from the StartKey via
+keys.RangeDescriptorKey, which in particular resides on the Range itself)
+that is accessible to the transactional KV API much like any other, but is
+for internal use only (and in particular plays no role at the SQL layer).
+
+Splits, Merges, and Rebalances are all carried out as distributed
+transactions. They are all complex in their own right but they share the
+basic approach of transactionally acting on the RangeDescriptor. Each of
+these operations at some point will
+
+- start a transaction
+
+- update the RangeDescriptor (for example, to reflect a split, or a change
+to the Replicas comprising the members of the Range)
+
+- update the meta ranges (which form a search index used for request routing)
+
+- commit with a roachpb.InternalCommitTrigger.
+
+In particular, note that the RangeDescriptor is the first write issued in the
+transaction, and so the transaction record will be created on the affected
+range. This allows us to establish a helpful invariant:
+
+INVARIANT: an intent on keys.RangeDescriptorKey is resolved atomically with
+the (application of the) roachpb.EndTxnRequest committing the transaction.
+
+To understand this invariant, it's best to try to make do without it and fail.
+Assume that a descriptor change committed, but that the intent will only get
+resolved asynchronously.
+
+- since the transaction committed, its update to the meta ranges becomes
+visible and the cluster will start routing requests according to the new
+descriptor. However, the Replicas may yet have to learn that the new
+descriptor is committed. This is not a correctness problem, but it is
+undesirable. The invariant avoids this, since the first intent to be resolved
+will be the RangeDescriptor on the leaseholder (during EndTxn application),
+and this will causally precede resolution of auxiliary intents since those
+will only be resolved after the leaseholder has applied the EndTxn.
+
+- each Replica (and its surrounding Store) needs to run a considerable amount
+of logic to update its state when switching to the new descriptor, and now
+this logic has to be triggered by the intent resolution. This is a lot less
```

Comment:
Why would the logic have to be done during intent resolution? Because we'd get infinite recursion if we tried to read it, since that would require the new range descriptor?

Reply:
Well, you switch to the new descriptor (atomically to keep disk and in-mem state in sync) and when it becomes visible, right? So it's when the intent gets removed. We make sure this always happens in the context of the EndTxn. If we don't do that, the ResolveIntent has to do it.

Reply:
But the descriptor is visible once the txn commits, not when the intent gets resolved, right? Assuming consistent MVCCful reads.

Reply:
Maybe we're talking past each other, but the descriptor with an intent on it is not visible. Readers will need to resolve the intent first (for all they know, it may be aborted). But also, we need to decide which version of the descriptor to use in `(*Store).Start`. If the txn record isn't local to the store, how do you know whether the intent or the last committed version is the one to use? It needs to be a deterministic local decision. So if there is an intent, you *have* to ignore it.

Reply:
I'm probably just not understanding the mechanics involved, seems worth reading up on. I suppose my point was that the text doesn't make it entirely clear why this is a problem, at least not to me, but it does seem nice to avoid the additional complexity of intent resolution.

> If the txn record isn't local to the store, how do you know whether the intent or the last committed version is the one to use?

I thought we always anchored these transactions on the same range as the range descriptor?

Reply:
Right, you could have this invariant:

> Changes to the range descriptor are always anchored on the same range and these txns cannot use parallel commit.

And then whenever you need to read the descriptor in, say, `(*Store).Start`, you could have custom code that goes to check the txn record in case of an intent.

but this is really just the old invariant, implemented in a more clumsy way - by checking the txn record you're checking whether the EndTxn has already applied. I was entertaining the case in which the txn record is *not* necessary local. Then applying the ResolveIntent is the only way for you to flip over.

Reply:
I agree, just resolving it during EndTxn definitely seems worthwhile here.